### PR TITLE
Core: Enhance PartitionsTable.Partition

### LINKED ISSUE #8455: Introduce PartitionEntry class to represent stats per partition
### Feature Request / Improvement

~In the design, The schema of partitions metadata table and partition stats file is kept to be same~

~Hence, move `PartitionsTable.Partition` to a separate class and re-use them for partition stats.
We also need to extend `PartitionsTable.Partition` this class to `org.apache.avro.generic.IndexedRecord` to make this class compatible to directly write and read from `ParquetAvroWriter` and `ParquetValueReaders`~

Introduce `PartitionEntry` class to represent stats per partition. This should be independent of `PartitionsTable`

### Query engine

None

### PR DESCRIPTION
**Pull request by ajantha-bhat**
`PartitionsTable.Partition` will be used between Partitions metadata table and partition stats reader-writer.
Hence, move it to a separate class and extend it with Avro's `IndexedRecord` (for partition stats writing).

Derived from #8488 

Fixes #8455

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/Partition.java [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        Types.NestedField.optional(10, Column.LAST_UPDATED_AT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            11, Column.LAST_UPDATED_SNAPSHOT_ID.name(), Types.LongType.get()));
+  }
+
+  private static Schema prepareAvroSchema(Types.StructType partitionType) {
+    return AvroSchemaUtil.convert(icebergSchema(partitionType), "partitionEntry");
+  }
+
+  public static class PartitionBuilder {
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/Partition.java:309]: I think we can just call this `Builder` (aligns with the naming convention we have for `TableMetadata`, `PartitionSpec`, `SnapshotRef` etc.

File: core/src/main/java/org/apache/iceberg/Partition.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+  public Long lastUpdatedAt() {
+    return lastUpdatedAt;
+  }
+
+  public Long lastUpdatedSnapshotId() {
+    return lastUpdatedSnapshotId;
+  }
+
+  synchronized void update(ContentFile<?> file, Snapshot snapshot) {
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/Partition.java:119]: Nit: why the change to make it synchronized, are we expecting to update the same Partition reference from multiple threads?

**Review by amogh-jahagirdar — COMMENTED** [review]
Thanks @ajantha-bhat very exciting to see all the work being done on this front! Just had a few comments/questions.

**Comment by ajantha-bhat** [comment]
> Also if I'm not mistaken the major pending aspect spec-wise is a conclusion on whether we maintain accurate record counts as opposed to eq/position delete counts right?

Yes, that we will conclude soon and get things moving.

File: core/src/main/java/org/apache/iceberg/Partition.java [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+ */
+package org.apache.iceberg;
+
+import java.util.Objects;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.types.Types;
+
+public class Partition implements IndexedRecord {
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/Partition.java:27]: If you see the PR https://github.com/apache/iceberg/pull/8503, partition stats reader and writer in the Util directly work with `Partition` as input and output. 

So, when the query engine wants to utilize partition stats for query planning, partition stats reader gives an iterator of `Partition`. Hence, this has to be public for engine integration module to utilize these stats.

File: core/src/main/java/org/apache/iceberg/Partition.java (hunk 2 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/Partition.java:119]: It depends on engine integration. Some engine like Dremio, writes manifest files and datafiles concurrently for a single snapshot operations. In that case we need a synchronization as there can be multiple data files belong to same partition.

File: core/src/main/java/org/apache/iceberg/Partition.java (hunk 1 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/Partition.java]: ACK. Updated.

File: core/src/main/java/org/apache/iceberg/Partition.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/Partition.java:27]: I see, in the end engine integrations would use `Partition` so it has to be public.

File: core/src/main/java/org/apache/iceberg/Partition.java (hunk 2 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/Partition.java:119]: Sure, I guess I was thinking that only after writing the data files (even if the data files are written concurrently) would we update the partition stats (serially) but as you said this is really an engine implementation detail.
